### DISCUSSION TITLE
Implementing socket.io with Node Cluster - error ERR_INTERNAL_ASSERTION when connected from Client

### DISCUSSION THREAD

**Discussion by chandregowda** [nodejs/help#4713] [Q&A]
### Version

v16.17.0

### Platform

Microsoft Windows NT 10.0.22000.0 x64

### What steps will reproduce the bug?

Node server.js - Once it sends the requested data to the connected socket  (from client side) and exit with the Exception 'ERR_INTERNAL_ASSERTION'

[server.zip](https://github.com/nodejs/node/files/9689363/server.zip)

### How often does it reproduce? Is there a required condition?

Every time I connect or reconnect to the Node server with socket from the Client, it is exiting with this error

### What is the expected behavior?

Socket connection should be established when running  node server with cluster mode

### What do you see instead?

```
node:internal/assert:14
    throw new ERR_INTERNAL_ASSERTION(message);
    ^

Error [ERR_INTERNAL_ASSERTION]: This is caused by either a bug in Node.js or incorrect usage of Node.js internals.
Please open an issue with this stack trace at https://github.com/nodejs/node/issues

    at new NodeError (node:internal/errors:387:5)
    at assert (node:internal/assert:14:11)
    at ServerResponse.detachSocket (node:_http_server:272:3)
    at resOnFinish (node:_http_server:847:7)
    at ServerResponse.emit (node:events:513:28)
    at onFinish (node:_http_outgoing:909:10)
    at afterWrite (node:internal/streams/writable:499:5)
    at afterWriteTick (node:internal/streams/writable:486:10)
    at processTicksAndRejections (node:internal/process/task_queues:82:21) {
  code: 'ERR_INTERNAL_ASSERTION'
}
```

### Additional information

Without Cluster implementation - it is working fine

**Comment by chandregowda** [comment]
Referred implementation here [https://socket.io/docs/v4/cluster-adapter/](url)

**Comment by bnoordhuis** [comment]
Can you post a minimal test case using only built-in modules? I.e., not express, socket.io, etc.

**Comment by ywave620** [comment]
@chandregowda This scipt included in the zip is not self-contained, could you please upload the `package.json` to describe the dependencies involved?  

> Referred implementation here [https://socket.io/docs/v4/cluster-adapter/](url)

I tried, this work fine in latest node

**Comment by Y0ursTruly** [comment]
I was using [canvas from nodejs](https://www.npmjs.com/package/canvas) and got the *"ERR_INTERNAL_ASSERATION"*
![image](https://user-images.githubusercontent.com/60293767/193702591-0ac5a235-da8c-4115-8104-c930bf358180.png)
If it means anything, I opened up command prompt in windows 10 and was in the nodejs command line where I pasted code before to get the `x[0]`

**Comment by chandregowda** [comment]
> @chandregowda This scipt included in the zip is not self-contained, could you please upload the `package.json` to describe the dependencies involved?
> 
> > Referred implementation here [https://socket.io/docs/v4/cluster-adapter/](url)
> 
> I tried, this work fine in latest node

package.json content

```
{
  "name": "nse",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "dev": "set PORT=80 && nodemon --ignore ./DATA/ --ignore ./tmp/ --ignore ./cloud/DATA/ --ignore ./rnd/ server.js",
    "start": "set PORT=80 && node server.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@socket.io/cluster-adapter": "^0.2.0",
    "@socket.io/sticky": "^1.0.1",
    "axios": "^0.26.1",
    "compression": "^1.7.4",
    "connect-history-api-fallback": "^2.0.0",
    "cors": "^2.8.5",
    "crypto-js": "^4.1.1",
    "express": "^4.17.3",
    "mongoose": "^6.4.4",
    "socket.io": "^4.5.2",
    "ws": "^7.4.0"
  },
  "engines": {
    "node": "16.x"
  },
  "devDependencies": {
    "nodemon": "^2.0.19"
  }
}

```

**Reply by chandregowda** [reply]
What is the node version you are using?

**Comment by chandregowda** [comment]
> Can you post a minimal test case using only built-in modules? I.e., not express, socket.io, etc.

My requirement is to use socket.io with cluster, and express is used only for routes and static folder handling

**Comment by bnoordhuis** [comment]
Then I'll go ahead and convert this to a discussion. So far there's no reason to believe it's a node bug vs. simply holding it wrong.

**Reply by chandregowda** [reply]
Can you please help me why it throws error after connecting to request from client socket.

**Reply by bnoordhuis** [reply]
I'm not going to sift through dozens of dependencies and thousands of lines of code. It's why I asked you to put together a minimal test case.

**Comment by konrad-g** [comment]
I also got this issue. In my case the problem was caused by the fact that I created socket.io connection with server already used by Express.js. Solution was to create a new http server for socket.io only. i.e.
```
import { Server } from "socket.io";

const httpServer = http.createServer();
const socketIo = new Server(httpServer);
```

**Reply by chandregowda** [reply]
Can you please share the Server code with Cluster (master and child) along with Express and Socket io
